### ISSUE TITLE
How should non-URN `id`s be encoded in `/things/{id}` URLs and `Location` headers?

### ISSUE THREAD

**Issue by benfrancis** [w3c/wot-discovery#382]
The Thing Description specification says that an id can be any URI. The examples in the Thing Discovery specification are all URNs, but that's only one type of URI.

From reading the specification, my assumption is that if you want to create a Thing Description with an HTTP URL as its ID, you'd have to submit an HTTP request like:

`PUT http://localhost:3000/things/https://mywebthingserver.com/things/lamp`

That looks a little strange, but it is a valid URL as far as I can tell. If you try that with [Zion](https://github.com/vaimee/zion) you get a 404 error:

```
{
    "statusCode": 404,
    "message": "Cannot PUT /things/https://mywebthingserver.com/things/lamp",
    "error": "Not Found"
}
```

Perhaps URLs as IDs like that should be URL encoded? E.g.

`PUT http://localhost:3000/things//https%3A%2F%2Fmywebthingserver.com%2Fthings%2Flamp`

Zion will accept that request, but you can end up with `http://localhost:3000/things/https%3A%2F%2Fmywebthingserver.com%2Fthings%2Flamp42` as the URL of the Thing Description and `https://mywebthingserver.com/things/lamp` as the `id` member inside the Thing Description. It actually doesn't seem to check that they are the same.

I've been wondering for a while how this would work in practice. Should the server accept `PUT /things/https://mywebthingserver.com/things/lamp`, or should it be URL encoded as `PUT /things/https%3A%2F%2Fmywebthingserver.com%2Fthings%2Flamp`?

If it's the former, then maybe an example would help illustrate that?

If it's the latter, then maybe some text is needed in the specification to deal with this?

**Comment by benfrancis** [comment]
For the record, currently if you add a web thing with an HTTP URL as an ID to WebThings Gateway, it gets given a URL like:

`https://user.webthings.io/things/https---mywebthingserver.com-things-lamp` (with all the non-alphanumeric characters replaced with dashes)

Since WebThings Gateway acts as a gateway, not just a directory, it also generates a new Thing Description for that Thing, using that URL as the new `id` member.

**Comment by relu91** [comment]
Good point, my take on this issue was that people would encode IDs that contain problematic characters, even though usually web devs are used to encode query parameters, not paths. for example I think postman autoenconding feature would break if you try to issue something like `http://localhost:3000/things/https://mywebthingserver.com/things/lamp`

>  then maybe some text is needed in the specification to deal with this?

+1

>https://user.webthings.io/things/https---mywebthingserver.com-things-lamp (with all the non-alphanumeric characters replaced with dashes)

Just curious, any reason why to choose dashes instead of URL encoding?

>Zion will accept that request, but you can end up with http://localhost:3000/things/https%3A%2F%2Fmywebthingserver.com%2Fthings%2Flamp42 as the URL of the Thing Description and https://mywebthingserver.com/things/lamp as the id member inside the Thing Description. It actually doesn't seem to check that they are the same.

I know this is kind of bothering but not a big issue IMO. The two strings are the same they are just encoded differently.

**Comment by farshidtz** [comment]
I believe this affects all the following:
```
PUT /things/{id}
GET /things/{id}
PATCH /things/{id}
DELETE /things/{id}
```
and the Location header value in response to `POST /things/{id}`.

**Comment by benfrancis** [comment]
**tl;dr:** Yeah, those IDs should probably be percent-encoded. Maybe even URNs too.

@relu91 wrote:
> Good point, my take on this issue was that people would encode IDs that contain problematic characters, even though usually web devs are used to encode query parameters, not paths. for example I think postman autoenconding feature would break if you try to issue something like `http://localhost:3000/things/https://mywebthingserver.com/things/lamp`

I did some reading...

It actually seems quite common for un-encoded URLs to be used as _query strings_, e.g.

`http://www.example.com?next=http://www.example.com`

RFC 3986 even [points out](https://datatracker.ietf.org/doc/html/rfc3986#section-3.4) that it may be more usable  _not_ to percent-encode them.

However, the `{id}` in `/things/{id}` is a path segment. The relevant parts of the ABNF grammar for a [path segment](https://www.rfc-editor.org/rfc/rfc3986#section-3.3) from RFC 3986 are:

`segment       = *pchar`
...
`pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"`
...
`unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"`
...
`pct-encoded   = "%" HEXDIG HEXDIG`
...
```
sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
                 / "*" / "+" / "," / ";" / "="
```

I don't have a lot of experience with these grammars, but my understanding of that is...

`http://localhost:3000/things/https://mywebthingserver.com/things/lamp?foo=bar#baz`

...would in fact be a valid URI, but may not behave as intended because `/`, `?` and `#` are reserved characters for delimiting path segments, query strings, and fragments respectively, so wouldn't be interpreted as part of the ID.

`gen-delims  = ":" / "/" / "?" / "#" / "[" / "]" / "@"`

A quick test with an Express.js router for the pattern `/:thingId` seems to show that a URL like...
`
http://localhost:8001/things/https://example.com/things/7670aa40-9802-4a01-a85a-7f853f81512b`

...would generate an error, at least in its default configuration.

In conclusion, although the original example provided would actually be a valid URI, the ID may not be parsed as intended. Therefore it's probably safest to percent-encode all URIs.

But does that mean we should percent encode URNs too? I think all of the URNs used as examples in the specification would technically be valid URI path segments, but I don't know if that applies to all URNs...

> Just curious, any reason why to choose dashes instead of URL encoding?

I don't know, I didn't make that decision. My guess is that it's because it's more human-readable, but I do now wonder if it risks causing duplicate IDs.

**Comment by benfrancis** [comment]
> I believe this affects all the following
<snip>

Good point, I've renamed the issue.

**Comment by benfrancis** [comment]
This is making what other things might need percent-encoding, like the query strings in the search API for example. I did a similar test with Express.js and an un-encoded JSONPath query was rejected as a query string, whereas a percent-encoded string seems to work. I don't know if that's an issue specific to Express.js, or whether it would be a widespread problem.

**Comment by mmccool** [comment]
@farshidtz has agreed to draft a PR to resolve this issue by adding a recommendation (assertion) to percent-encode an ID as a URI when necessary.  This also will apply to all other URI's, e.g. queries.

**Comment by mmccool** [comment]
Seems related to issue #320 which is also asking about URL-encoding.  So the solution to this issue should ideally close that one, too.

**Comment by benfrancis** [comment]
Just to report back on this following implementation experience, I've found that for some URL routers (like [Express.js](https://expressjs.com/)) encoding a URI with `encodeURIComponent()` in JavaScript is sufficient, but for others (like [page.js](https://visionmedia.github.io/page.js/)) you actually need to double encode them in order for URLs to match a pattern like `/things/{id}`, since the router does one pass of URI decoding as part of its operation. This is likely a bug in the URL router, but is a potential gotcha for implementers.
